lib/util/pthread_create_blocking_np.c

```diff
@@ -0,0 +1,170 @@
+#include <sys/time.h>
+
+#include <errno.h>
+#include <pthread.h>
+#include <stdlib.h>
+
+#include "pthread_create_blocking_np.h"
+
+struct wrapped_cookie {
+	/* User-supplied. */
+	void * (*start_routine)(void *);
+	void * arg;
+
+	/* Synchronization. */
+	pthread_mutex_t mutex;
+	pthread_cond_t cond;
+	int running;		/* 1 if the thread has started. */
+
+	/* In case the synchronization fails. */
+	int rc_sync;		/* non-zero if synchronization failed. */
+};
+
+/* Set the "running" flag and signal the condition variable. */
+static int
+set_running(struct wrapped_cookie * U)
+{
+	int rc;
+
+	/* Lock mutex. */
+	if ((rc = pthread_mutex_lock(&U->mutex)))
+		goto err0;
+
+	/* Set the "running" flag and signal the condition variable. */
+	U->running = 1;
+	if ((rc = pthread_cond_signal(&U->cond)))
+		goto err1;
+
+	/* Unlock mutex. */
+	if ((rc = pthread_mutex_unlock(&U->mutex)))
+		goto err0;
+
+	/* Success! */
+	return (0);
+
+err1:
+	/*
+	 * Don't record any errors in this clean-up; the existing value of rc
+	 * is more important.
+	 */
+	pthread_mutex_unlock(&U->mutex);
+err0:
+	/* Failure! */
+	return (rc);
+}
+
+/* Routine which is executed by pthread_create(). */
+static void *
+wrapped_thread(void * cookie)
+{
+	struct wrapped_cookie * U = cookie;
+	void * (*start_routine)(void *);
+	void * arg;
+	int rc;
+
+	/* Copy the user-supplied parameters, because U will not be valid. */
+	start_routine = U->start_routine;
+	arg = U->arg;
+
+	/* Indicate that the main thread can continue (which will free U). */
+	if ((rc = set_running(U)) != 0) {
+		/* If there was an error, then we can still write to U. */
+		U->rc_sync = rc;
+		return (NULL);
+	}
+
+	/* Run the desired routine. */
+	return (start_routine(arg));
+}
+
+/**
+ * pthread_create_blocking_np(thread, attr, start_routine, arg):
+ * Run pthread_create() and block until the the ${thread} has started.  The
+ * thread will execute ${start_routine} with ${arg} as its sole argument.
+ * When ${start_routine} finishes, make its returned value available via
+ * pthread_join().  If successful, return 0; otherwise return the error number.
+ */
+int
+pthread_create_blocking_np(pthread_t * restrict thread,
+    const pthread_attr_t * restrict attr,
+    void * (*start_routine)(void *), void * restrict arg)
+{
+	struct wrapped_cookie * U;
+	int rc;
+
+	/* Allocate our cookie and store parameters. */
```

Comment:
Worth a comment here mentioning that the C standard does not require that
variables with automatic storage duration are accessible by other threads;
POSIX (and thus pthreads) does provide that guarantee, but out of an abundance
of caution we take the option of using malloc in case this code ends up
running on a non-POSIX system.